scala/scala-impl/src/org/jetbrains/plugins/scala/lang/refactoring/move/anonymousToInner/ScalaAnonymousToInnerHandler.scala

```diff
@@ -0,0 +1,157 @@
+            .add(newClassFromAnonymous)
+
+          maybeWarningComment.foreach(comment => containerWithNewInnerClass.addBefore(comment, newClassFromAnonymous))
+          originalElement.replace(innerClassNewTemplate)
+        }
+        catch {
+          case e: IncorrectOperationException =>
+            LOG.error(e)
+        }
+
+      }
+      ApplicationManager.getApplication.runWriteAction(action)
+
+    }, getRefactoringName, null)
+
+  private def newTemplateForInnerClass(project: Project, name: String, variables: Array[ScalaVariableData], newTemplate: ScNewTemplateDefinition) = {
+    implicit val projectContext: ProjectContext = new ProjectContext(project)
+
+    val args = variables.map(_.variable.getName).mkString(", ")
+    val text = s"""new $name($args)""".stripMargin
+
+    createElementFromText[ScNewTemplateDefinition](text, newTemplate)
+  }
+
+  private def collectUsedVariables(anonClass: ScExtendsBlock, targetContainer: Either[ScFile, ScTemplateDefinition]): Array[ScalaVariableData] = {
+    var res: List[ScTypedDefinition] = Nil
+    anonClass.accept(new ScalaRecursiveElementVisitor() {
+      override def visitReferenceExpression(expression: ScReferenceExpression) = {
+        val refElement = expression.resolve
+        refElement match {
+          case p: ScParameter =>
+            val containingClass = p.getDeclarationScope
+            if (targetContainer.fold(_ => false, scTemplateDefinition => containingClass != scTemplateDefinition && !containingClass.isAncestorOf(scTemplateDefinition)))
+              res = p :: res
+          case r: ScReferencePattern =>
+            val containingClass = r.containingClass
+            if (targetContainer.fold(_ => false, scTemplateDefinition => containingClass != scTemplateDefinition && !containingClass.isAncestorOf(scTemplateDefinition) && !findTargetContainer(r).isLeft))
+              res = r :: res
+          case _ =>
+        }
+        super.visitReferenceExpression(expression)
+      }
+    })
+    res.reverse.distinct.map(parameter =>
+      new ScalaVariableData(parameter, true, parameter.`type`().getOrNothing)
+    ).toArray
+
+  }
+
+  private case class DialogResult(className: String, variables: Array[ScalaVariableData])
+
+  private def showRefactoringDialog(project: Project, extendsBlock: ScExtendsBlock, usedVariables: Array[ScalaVariableData], target: Either[ScFile, ScTemplateDefinition]) = {
+    val dialog = new ScalaAnonymousToInnerDialog(project, extendsBlock, usedVariables, target)
+    if (dialog.showAndGet())
+      Some(DialogResult(dialog.getClassName, dialog.getVariables))
+    else
+      None
+  }
+
+  private def createClass(name: String, anonClass: ScExtendsBlock, variables: Array[ScalaVariableData], project: Project): (ScClass, Option[PsiComment]) = {
+    implicit val projectContext: ProjectContext = new ProjectContext(project)
+
+    val containsVars = variables.exists(_.element.isVar)
+    val warningComment = if (containsVars)
```

Comment:
(I will add it as another comment to track it easier)
I've just realised that simply ignoring "write" operations and adding the comment is not enough.
Here is a modified example:
```scala
trait MyTrait {
  def foo(): Unit
}

object Example {
  def main(args: Array[String]): Unit = {
    var localVar = 0

    val t = new MyTrait {
      override def foo(): Unit = {
        println(localVar)
      }
    }

    t.foo()
    localVar += 1
    t.foo()
    localVar += 1

    println(localVar)
  }
}
```

Notice that `localVar` is not modified inside the anonymous class.
It's only accessed inside it.
But the variable is modified in some other places outside the anonymous class.

The code outputs:
```
0
1
2
```

When you extract the class it looks like this:
```scala
trait MyTrait {
  def foo(): Unit
}

object Example {
  def main(args: Array[String]): Unit = {
    var localVar = 0

    val t = new MyMyTrait(localVar)

    t.foo()
    localVar += 1
    t.foo()
    localVar += 1

    println(localVar)
  }
  // refactor your code or use scala.runtime.ObjectRef

  class MyMyTrait(localVar: Int) extends MyTrait {
    override def foo(): Unit = {
      println(localVar)
    }
  }
}
```

Now `MyMyTrait` accesses its local copy of the var.
And the output of the program is:
```
0
0
2
```

Reply:
It's a simple example, but it illustrates that the refactoring still can break the code silently.
I see two solutions:

1. Ignore `var`s, don't include them as a parameter.
The code can potentially become red, and user will have to handle it manually.
Leave `//todo` comments where needed to mention that we can't handle vars yet and user needs to use `scala.runtime.*Ref` objects manually.

2. Rewrite the code to `*Ref`
E.g. the example from above will look like:
```scala
import scala.runtime.IntRef

trait MyTrait {
  def foo(): Unit
}

object Example {
  def main(args: Array[String]): Unit = {
    val localVar = new IntRef(0)

    val t = new MyMyTrait(localVar)

    t.foo()
    localVar.elem += 1
    t.foo()
    localVar.elem += 1

    println(localVar)
  }

  class MyMyTrait(localVar: IntRef) extends MyTrait {
    override def foo(): Unit = {
      println(localVar.elem)
    }
  }
}
```

The first approach is the easiest.
If you choose it, please create a new YT ticket (and link to https://youtrack.jetbrains.com/issue/SCL-227) describing the limitation and mentioning the 2nd approach.

The second approach is the correct one, but would require some extra effort.
It might be practical to handle it as a separate change.
The refactoring without `var` handling would be already useful.
Some points to consider:
- Mind that there is `scala.runtime.ObjectRef` and multiple `*Ref` classes for primitive types.
- Mind that there are `scala.runtime.*Ref` and `scala.runtime.Volatile*Ref`

Reply:
In option 1 if we don't include the vars in the parameters they can accidentally reference another field with the same name in the target container and therefore compile without errors. extending your example:

```
trait MyTrait {
  def foo(): Unit
}

object Example {

  var localVar = 10
  
  def main(args: Array[String]): Unit = {
    var localVar = 0

    val t = new MyTrait {
      override def foo(): Unit = {
        println(localVar)
      }
    }

    t.foo()
    localVar += 1
    t.foo()
    localVar += 1

    println(localVar)
  }
}
```
that prints
```
0
1
2
```


would become:
```trait MyTrait {
  def foo(): Unit
}

object Example {

  var localVar = 10
  
  def main(args: Array[String]): Unit = {
    var localVar = 0

    val t = new MyMyTrait()

    t.foo()
    localVar += 1
    t.foo()
    localVar += 1

    println(localVar)
  }

  class MyMyTrait() extends MyTrait {
    override def foo(): Unit = {
      println(localVar) //todo: refactor your code or use scala.runtime.ObjectRef
    }
  }
}
```
will print
```
10
10
2
```

Reply:
> As a possible quick solution, I could just throw an error to the user if the code contains var references instead of opening the dialog

Indeed, a good tradeoff.
Let's do it.
We can show a red error tooltip to the user, like this:
<img width="489" alt="image" src="https://github.com/JetBrains/intellij-scala/assets/3989292/e33d46d8-d8b8-42fe-8c41-f0294f3dc251">
You might use `com.intellij.refactoring.util.CommonRefactoringUtil#showErrorHint`